### PR TITLE
ui: update public key form to choose filtering

### PR DESCRIPTION
**Pull request by noreng-jg**

### CODE REVIEW

File: api/routes/sshkeys.go [hunk 1]
```diff
@@ -170,10 +181,129 @@ func (h *Handler) EvaluateKey(c gateway.Context) error {
...
+	var tenant string
+	if c.Tenant() != nil {
+		tenant = c.Tenant().ID
+	}
+
+	err := guard.EvaluatePermission(c.Role(), authorizer.Actions.PublicKey.AddTag, func() error {
+		var err error
+		err = h.service.AddPublicKeyTag(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), req.Tag)
+
+		return err
```
Comment (otavio) [inline api/routes/sshkeys.go:209]:
```suggestion
		return h.service.AddPublicKeyTag(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), req.Tag)
```

File: api/routes/sshkeys.go [hunk 2]
```diff
@@ -170,10 +181,129 @@ func (h *Handler) EvaluateKey(c gateway.Context) error {
...
+	var tenant string
+	if c.Tenant() != nil {
+		tenant = c.Tenant().ID
+	}
+
+	err := guard.EvaluatePermission(c.Role(), authorizer.Actions.PublicKey.RemoveTag, func() error {
+		var err error
+		err = h.service.RemovePublicKeyTag(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), c.Param(ParamTagName))
+
+		return err
```
Comment (otavio) [inline api/routes/sshkeys.go:245]:
```suggestion
		return h.service.RemovePublicKeyTag(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), c.Param(ParamTagName))
```

File: api/routes/sshkeys.go [hunk 3]
```diff
@@ -170,10 +181,129 @@ func (h *Handler) EvaluateKey(c gateway.Context) error {
...
+	var tenant string
+	if c.Tenant() != nil {
+		tenant = c.Tenant().ID
+	}
+
+	err := guard.EvaluatePermission(c.Role(), authorizer.Actions.PublicKey.UpdateTag, func() error {
+		var err error
+		err = h.service.UpdatePublicKeyTags(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), req.Tags)
+
+		return err
```
Comment (otavio) [inline api/routes/sshkeys.go:285]:
```suggestion
		return h.service.UpdatePublicKeyTags(c.Ctx(), tenant, c.Param(ParamPublicKeyFingerprint), req.Tags)
```

File: api/services/sshkeys.go [hunk 4]
```diff
@@ -30,17 +30,35 @@ type Request struct {
...
+	if key.Filter.Hostname != "" {
+		ok, err := regexp.MatchString(key.Filter.Hostname, dev.Name)
+		if err != nil {
+			return false, err
+		}
+
+		return ok, nil
+	} else if len(key.Filter.Tags) > 0 {
+		for _, tag := range dev.Tags {
+			if exist(tag, key.Filter.Tags) {
```
Comment (otavio) [inline api/services/sshkeys.go:53]: I understood the use of lambda here however it is used only once; I'd put the `for` here.

File: api/services/sshkeys.go [hunk 5]
```diff
@@ -61,6 +79,39 @@ func (s *service) GetPublicKey(ctx context.Context, fingerprint, tenant string)
...
+	// The tags are valid when they exist in some device.
+	if key.Filter.Tags != nil {
+		if len(key.Filter.Tags) > 0 {
+			tags, _, err := s.store.DeviceGetTags(ctx, tenant)
+			if err != nil {
+				return err
+			}
+
+			for _, tag := range key.Filter.Tags {
+				if !exist(tag, tags) {
```
Comment (otavio) [inline api/services/sshkeys.go:106]: Same here. 

If desired to use a function (which I am fine too) I'd use https://pkg.go.dev/sort#SearchStrings.

File: api/services/sshkeys.go [hunk 6]
```diff
@@ -92,6 +143,39 @@ func (s *service) ListPublicKeys(ctx context.Context, pagination paginator.Query
...
+	// The tags are valid when they exist in some device.
+	if key.Filter.Tags != nil {
+		if len(key.Filter.Tags) > 0 {
+			tags, _, err := s.store.DeviceGetTags(ctx, tenant)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, tag := range key.Filter.Tags {
+				if !exist(tag, tags) {
```
Comment (otavio) [inline api/services/sshkeys.go:170]: Same here.

File: api/services/sshkeys_tags.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	if len(key.Filter.Tags) == DeviceMaxTags {
+		return ErrMaxTagReached
+	}
+
+	tags, _, err := s.GetTags(ctx, tenant)
+	if err != nil {
+		return err
+	}
+
+	if !exist(tag, tags) {
```
Comment (otavio) [inline api/services/sshkeys_tags.go:54]: Here too.

File: api/services/sshkeys_tags.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil || key == nil {
+		return ErrPublicKeyNotFound
+	}
+
+	if key.Filter.Hostname != "" {
+		return ErrPublicKeyInvalid
+	}
+
+	// Checks if the tag already exists in the device.
+	if !exist(tag, key.Filter.Tags) {
```
Comment (otavio) [inline api/services/sshkeys_tags.go:100]: And here.

File: api/store/mongo/device_store_tags.go [hunk 9]
```diff
@@ -50,18 +54,53 @@ func (s *Store) DeviceUpdateTag(ctx context.Context, uid models.UID, tags []stri
...
+			helper := make([]string, len(a))
+			for i, value := range list {
+				helper[i] = fmt.Sprint(value)
+			}
+
+			return helper
+		}
+
+		return append(convert(a), convert(b)...)
+	}
```
Comment (otavio) [inline api/store/mongo/device_store_tags.go:80]: Something like https://github.com/emirpasic/gods#sets could help.

**Review by otavio — CHANGES_REQUESTED** [review]

File: api/routes/sshkeys.go (hunk 1 above)
Comment (henrybarreto) [reply api/routes/sshkeys.go:209]: Solved on API PR.

https://github.com/shellhub-io/shellhub/pull/1711/files#diff-cfef764ad3b2b7db5e03ce88366635533209cca6671c99f668427b964de580f0R205-R209

**Review by otavio — APPROVED** [review]

**Pull request merged**